mypy/checker.py

```diff
@@ -488,16 +235,23 @@ def accept(self, node: Node, type_context: Type = None) -> Type:
         else:
             return typ
 
-    def accept_in_frame(self, node: Node, type_context: Type = None,
-                        repeat_till_fixed: bool = False) -> Type:
-        """Type check a node in the given type context in a new frame of inferred types."""
-        while True:
-            self.binder.push_frame()
-            answer = self.accept(node, type_context)
-            changed, _ = self.binder.pop_frame(True, True)
-            self.breaking_out = False
-            if not repeat_till_fixed or not changed:
-                return answer
+    def accept_loop(self, body: Node, else_body: Node = None) -> Type:
+        """Repeatedly type check a loop body until the frame doesn't change.
+
+        Then check the else_body.
+        """
+        # The outer frame accumulates the results of all iterations
+        with self.binder.frame_context(1) as outer_frame:
+            self.binder.push_loop_frame()
+            while True:
+                outer_frame.options_on_return.append(outer_frame)  # We may skip each iteration
+                with self.binder.frame_context(1) as inner_frame:
+                    self.accept(body)
+                if not inner_frame.changed:
+                    break
+            self.binder.pop_loop_frame()
```

Comment:
OK, now it (and broken) are placed at `binder.last_pop_changed` and `binder.last_pop_breaking_out`, and the Frame is just a dict.